herd/ASLSem.ml

```diff
@@ -435,13 +449,26 @@ module Make (C : Config) = struct
       | [] | [ _ ] | _ :: _ :: _ :: _ ->
           Warn.fatal "Arity error for function %s." name
 
+    let arity_three name args return_type f =
+      build_primitive name args return_type @@ function
+      | [ x; y; z; ] -> f x y z
+      | _ ->
+          Warn.fatal "Arity error for function %s." name
+
+    let arity_four name args return_type f =
+      build_primitive name args return_type @@ function
+      | [ x; y; z; t;] -> f x y z t
+      | _ ->
+          Warn.fatal "Arity error for function %s." name
+
     let return_one ty = (Some ty, fun body args -> return [ body args ])
     let return_zero = (None, Fun.id)
 
     (* Primitives *)
     let extra_funcs ii_env =
       let open AST in
       let with_pos = Asllib.ASTUtils.add_dummy_pos in
+      let bool = T_Bool |> with_pos in
```

Comment:
Can I suggest the name `boolean`? That's how it's written in ASL

However `d` just under could be renamed to `integer` also, and `bv_64` to `bits_64` to be coherent so maybe for another time ...